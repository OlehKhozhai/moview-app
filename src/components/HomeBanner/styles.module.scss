@import 'assets/styles/mixins/colors';

.root {
  position: relative;
  padding: 110px 120px 160px;

  &::before {
    content: '';
    background: center / cover no-repeat url('assets/images/home-banner-bg.jpg');
    position: absolute;
    top: 10px;
    bottom: 10px;
    right: 10px;
    left: 10px;
    filter: blur(10px);
  }
}

.header {
  display: flex;
  align-items: center;
  justify-content: space-between;
  position: absolute;
  top: 20px;
  left: 60px;
  right: 60px;
}

.title {
  text-transform: uppercase;
  font-size: 40px;
  margin-bottom: 40px;
  color: $primary-color;
}
