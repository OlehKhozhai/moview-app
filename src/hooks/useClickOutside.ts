import { RefObject, useEffect, useState } from 'react';

type Props = {
  ref: RefObject<HTMLElement>;
  handler?: keyof GlobalEventHandlersEventMap;
  closeOnOutsideClick?: boolean;
  callback?: () => void;
};

const useClickOutside = ({
  ref,
  callback,
  handler = 'click',
  closeOnOutsideClick = true,
}: Props) => {
  const [isClickOutside, setClickOutside] = useState(false);

  useEffect(() => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const handleClick = (e: any) => {
      const targetIsValid = e.target && document.contains(e.target);
      if (ref && ref.current && targetIsValid) {
        const isClickOnRef = ref.current.contains(e.target);

        if (!isClickOnRef && closeOnOutsideClick) {
          setClickOutside(true);
          if (callback) {
            callback();
          }
        }
      }
    };

    if (ref?.current) {
      document.addEventListener(handler, handleClick);
    }
    return () => {
      document.removeEventListener(handler, handleClick);
    };
  }, [handler, ref, closeOnOutsideClick, callback]);

  return [isClickOutside];
};

export default useClickOutside;
