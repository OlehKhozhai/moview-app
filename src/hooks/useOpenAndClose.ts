import React from 'react';

type Props = { onCloseCallback?: () => void } | undefined;

const useOpenAndClose = (props: Props = {}) => {
  const { onCloseCallback } = props;
  const [isOpen, setIsOpen] = React.useState(false);

  const handleClose = React.useCallback(() => {
    if (onCloseCallback) {
      onCloseCallback();
    }

    setIsOpen(false);
  }, [onCloseCallback]);

  const handleOpen = React.useCallback(() => {
    setIsOpen(true);
  }, []);

  const handleToggle = React.useCallback(
    (e?: React.MouseEvent<HTMLElement>) => {
      if (e) {
        e.stopPropagation();
      }

      setIsOpen((prevState) => {
        if (!prevState && onCloseCallback) {
          onCloseCallback();
        }

        return !prevState;
      });
    },
    [onCloseCallback]
  );

  return {
    isOpen,
    onClose: handleClose,
    onOpen: handleOpen,
    onToggle: handleToggle,
  };
};

export default useOpenAndClose;
