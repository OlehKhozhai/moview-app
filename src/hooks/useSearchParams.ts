import { useEffect, useState } from 'react';
import { useHistory } from 'react-router-dom';

import { DROPDOWN_OPTIONS } from 'config';

const useSearchParams = (): string => {
  const [params, setParams] = useState('');

  const {
    replace: historyReplace,
    location: { search: searchQuery },
  } = useHistory();

  useEffect(() => {
    const searchParams = new URLSearchParams(searchQuery);
    const activeTab = searchParams.get('search');
    const sortBy = searchParams.get('sortBy');

    if (!activeTab) {
      searchParams.set('searchBy', 'genres');
      searchParams.set('search', '');
    }

    if (!sortBy) {
      searchParams.set('sortBy', DROPDOWN_OPTIONS[0]);
    }

    historyReplace({ search: searchParams.toString() });
    setParams(searchParams.toString());
  }, [searchQuery, historyReplace]);

  return params;
};

export default useSearchParams;
