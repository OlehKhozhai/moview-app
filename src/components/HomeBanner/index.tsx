import React from 'react';
import cn from 'classnames';

import Search from 'components/Search';
import Button from 'components/_common/Button';
import Logo from 'components/_common/Logo';
import Modal from 'components/_common/Modal';
import useOpenAndClose from 'hooks/useOpenAndClose';
import CreateAndEditMovie from 'components/_common/CreateAndEditMovie';
import styles from './styles.module.scss';

type HomeBannerProps = {
  className?: string;
};

const HomeBanner: React.FC<HomeBannerProps> = ({ className }) => {
  const { isOpen, onClose, onOpen } = useOpenAndClose({});

  return (
    <>
      <div className={cn(styles.root, className)}>
        <div className={styles.header}>
          <Logo />
          <Button title="+ add movie" variation="secondary" onClick={onOpen} />
        </div>
        <h2 className={styles.title}>Find your movie</h2>
        <Search />
      </div>

      {isOpen && (
        <Modal isOpen={isOpen} onClose={onClose}>
          <CreateAndEditMovie />
        </Modal>
      )}
    </>
  );
};

export default React.memo(HomeBanner);
