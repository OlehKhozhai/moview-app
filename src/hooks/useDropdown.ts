import React from 'react';

import useOpenAndClose from 'hooks/useOpenAndClose';

type Props =
  | {
      defaultValue?: string;
    }
  | undefined;

const useDropdown = (props: Props = {}) => {
  const { defaultValue } = props;
  const [value, setValue] = React.useState(defaultValue || '');
  const { isOpen, onToggle, onClose } = useOpenAndClose();

  const handleOptionClick = React.useCallback(
    (option: string) => {
      setValue(option);
      onClose();
    },
    [onClose]
  );

  return {
    value,
    onOptionClick: handleOptionClick,
    isDropdownVisible: isOpen,
    onToggleDropdownVisibility: onToggle,
    onCloseDropdown: onClose,
  };
};

export default useDropdown;
