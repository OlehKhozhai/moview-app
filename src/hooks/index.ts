import useOpenAndClose from './useOpenAndClose';
import useDropdown from './useDropdown';
import useClickOutside from './useClickOutside';

export default { useOpenAndClose, useDropdown, useClickOutside };
