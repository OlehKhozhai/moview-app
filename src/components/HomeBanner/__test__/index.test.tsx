import React from 'react';
import { create } from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';

import hooks from 'hooks';
import HomeBanner from '../index';

jest.mock('components/_common/Modal', () => ({
  __esModule: true,
  default: () => '---Modal---',
}));

describe('HomeBanner component', () => {
  test('should render HomeBanner component by default', () => {
    const component = create(
      <MemoryRouter initialEntries={['/']}>
        <HomeBanner />
      </MemoryRouter>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('should render HomeBanner component with opened modal', () => {
    jest.spyOn(hooks, 'useOpenAndClose').mockReturnValue({
      isOpen: true,
      onOpen: jest.fn(),
      onClose: jest.fn(),
      onToggle: jest.fn(),
    });

    const component = create(
      <MemoryRouter initialEntries={['/']}>
        <HomeBanner />
      </MemoryRouter>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
