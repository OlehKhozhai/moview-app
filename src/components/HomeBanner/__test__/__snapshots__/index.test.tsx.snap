// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeBanner component should render HomeBanner component by default 1`] = `
<div
  className="root"
>
  <div
    className="header"
  >
    <a
      className="root"
      href="/"
      onClick={[Function]}
    >
      <span
        className="bold"
      >
        netflix
      </span>
      roulete
    </a>
    <button
      className="root secondary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      + add movie
    </button>
  </div>
  <h2
    className="title"
  >
    Find your movie
  </h2>
  <div
    className="root"
  >
    <input
      className="input"
      onChange={[Function]}
      placeholder="What do you want to watch?"
      value=""
    />
    <button
      className="root primary button"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Search
    </button>
  </div>
</div>
`;

exports[`HomeBanner component should render HomeBanner component with opened modal 1`] = `
Array [
  <div
    className="root"
  >
    <div
      className="header"
    >
      <a
        className="root"
        href="/"
        onClick={[Function]}
      >
        <span
          className="bold"
        >
          netflix
        </span>
        roulete
      </a>
      <button
        className="root secondary"
        disabled={false}
        onClick={[MockFunction]}
        type="button"
      >
        + add movie
      </button>
    </div>
    <h2
      className="title"
    >
      Find your movie
    </h2>
    <div
      className="root"
    >
      <input
        className="input"
        onChange={[Function]}
        placeholder="What do you want to watch?"
        value=""
      />
      <button
        className="root primary button"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Search
      </button>
    </div>
  </div>,
  "---Modal---",
]
`;
